chronos/streams/tlsstream.nim

```diff
@@ -0,0 +1,280 @@
+  var rstream = cast[TlsStreamReader](stream)
+  var wstream = rstream.stream.writer
+  var engine: ptr SslEngineContext
+  if rstream.kind == TlsStreamKind.Server:
+    engine = addr rstream.scontext.eng
+  else:
+    engine = addr rstream.ccontext.eng
+
+  try:
+    var length: uint
+    while true:
+      var state = engine.sslEngineCurrentState()
+      if (state and SSL_CLOSED) == SSL_CLOSED:
+        let err = engine.sslEngineLastError()
+        if err != 0:
+          rstream.error = newTlsStreamProtocolError(err)
+          rstream.state = AsyncStreamState.Error
+          break
+        else:
+          rstream.state = AsyncStreamState.Stopped
+          break
+
+      if (state and SSL_SENDREC) == SSL_SENDREC:
+        # TLS record needs to be sent over stream.
+        length = 0'u
+        var buf = sslEngineSendrecBuf(engine, length)
+        doAssert(length != 0 and not isNil(buf))
+        await wstream.wsource.write(buf, int(length))
+        sslEngineSendrecAck(engine, length)
+        continue
+
+      if (state and SSL_SENDAPP) == SSL_SENDAPP:
+        # Application data can be sent over stream.
+        if len(wstream.queue) > 0:
+          var item = await wstream.queue.get()
+          if item.size > 0:
+            length = 0'u
+            var buf = sslEngineSendappBuf(engine, length)
+            let toWrite = min(int(length), item.size)
+
+            if int(length) >= item.size:
+              if item.kind == Pointer:
+                let p = cast[pointer](cast[uint](item.data1) + uint(item.offset))
+                copyMem(buf, p, item.size)
+              elif item.kind == Sequence:
+                copyMem(buf, addr item.data2[item.offset], item.size)
+              elif item.kind == String:
+                copyMem(buf, addr item.data3[item.offset], item.size)
+              sslEngineSendappAck(engine, uint(item.size))
+              sslEngineFlush(engine, 0)
+              item.future.complete()
+            else:
+              if item.kind == Pointer:
+                let p = cast[pointer](cast[uint](item.data1) + uint(item.offset))
+                copyMem(buf, p, length)
+              elif item.kind == Sequence:
+                copyMem(buf, addr item.data2[item.offset], length)
+              elif item.kind == String:
+                copyMem(buf, addr item.data3[item.offset], length)
+              item.offset = item.offset + int(length)
+              item.size = item.size - int(length)
+              wstream.queue.addFirstNoWait(item)
+              sslEngineSendappAck(engine, length)
+            continue
+          else:
+            # Zero length item means finish
+            rstream.state = AsyncStreamState.Finished
+            break
+
+      if (state and SSL_RECVREC) == SSL_RECVREC:
+        # TLS records required for further processing
+        length = 0'u
+        var buf = sslEngineRecvrecBuf(engine, length)
+        let res = await rstream.rsource.readOnce(buf, int(length))
```

Comment:
This needs to be interrupted when write queue changes its size from 0 to 1. Otherwise we hang forever, while the server waits for us.